pkg/scheduler/ringmanager.go

```diff
@@ -0,0 +1,244 @@
+
+	if !cfg.UseSchedulerRing {
+		return nil, fmt.Errorf("ring manager shouldn't be invoked when ring is not used for discovering schedulers")
+	}
+
+	// instantiate kv store for both modes.
+	ringStore, err := kv.NewClient(
+		rm.cfg.SchedulerRing.KVStore,
+		ring.GetCodec(),
+		kv.RegistererWithKVName(prometheus.WrapRegistererWithPrefix("loki_", registerer), "scheduler"),
+		rm.log,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "scheduler ring manager create KV store client")
+	}
+
+	// instantiate ring for both mode modes.
+	ringCfg := rm.cfg.SchedulerRing.ToRingConfig(ringReplicationFactor)
+	rm.Ring, err = ring.NewWithStoreClientAndStrategy(ringCfg, ringNameForServer, ringKey, ringStore, ring.NewIgnoreUnhealthyInstancesReplicationStrategy(), prometheus.WrapRegistererWithPrefix("cortex_", registerer), rm.log)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create ring client for scheduler ring manager")
+	}
+
+	if managerMode == RingManagerModeMember {
+		if err := rm.startMemberMode(ringStore, registerer); err != nil {
+			return nil, err
+		}
+		return rm, nil
+	}
+
+	if err := rm.startReaderMode(); err != nil {
+		return nil, err
+	}
+	return rm, nil
+}
+
+func (rm *RingManager) startMemberMode(ringStore kv.Client, registerer prometheus.Registerer) error {
+	lifecyclerCfg, err := rm.cfg.SchedulerRing.ToLifecyclerConfig(ringNumTokens, rm.log)
+	if err != nil {
+		return errors.Wrap(err, "invalid ring lifecycler config")
+	}
+
+	delegate := ring.BasicLifecyclerDelegate(rm)
+	delegate = ring.NewLeaveOnStoppingDelegate(delegate, rm.log)
+	delegate = ring.NewTokensPersistencyDelegate(rm.cfg.SchedulerRing.TokensFilePath, ring.JOINING, delegate, rm.log)
+	delegate = ring.NewAutoForgetDelegate(ringAutoForgetUnhealthyPeriods*rm.cfg.SchedulerRing.HeartbeatTimeout, delegate, rm.log)
+
+	rm.RingLifecycler, err = ring.NewBasicLifecycler(lifecyclerCfg, ringNameForServer, ringKey, ringStore, delegate, rm.log, registerer)
+	if err != nil {
+		return errors.Wrap(err, "failed to create ring lifecycler for scheduler ring manager")
+	}
+
+	svcs := []services.Service{rm.RingLifecycler, rm.Ring}
+	rm.subservices, err = services.NewManager(svcs...)
+	if err != nil {
+		return errors.Wrap(err, "failed to create services manager for scheduler ring manager in member mode")
+	}
+
+	rm.subservicesWatcher = services.NewFailureWatcher()
+	rm.subservicesWatcher.WatchManager(rm.subservices)
+	rm.Service = services.NewBasicService(rm.starting, rm.running, rm.stopping)
+
+	return nil
+}
+
+func (rm *RingManager) startReaderMode() error {
+	var err error
+
+	svcs := []services.Service{rm.Ring}
+	rm.subservices, err = services.NewManager(svcs...)
+	if err != nil {
+		return errors.Wrap(err, "failed to create services manager for scheduler ring manager in reader mode")
+	}
+
+	rm.subservicesWatcher = services.NewFailureWatcher()
+	rm.subservicesWatcher.WatchManager(rm.subservices)
+
+	rm.Service = services.NewIdleService(func(ctx context.Context) error {
```

Comment:
sorry for the lack of understanding, but I'm not quite following how this IdleService works, and how it is able to read the ring without adding itself to it? Is it because it only has a `Ring` service and not a `RingLifecycler` service?

Reply:
When running `RingManager` in reader mode, we only use the ring client for reading the ring. We do not have to register the service in the ring to be able to read the ring. `RingLifecycler` is required only when we want to register tokens in the ring.